llvm/lib/Target/AIE/AIE2InstructionSelector.cpp

```diff
@@ -4140,6 +4266,165 @@ bool AIE2InstructionSelector::select512BitG_AIE_STORE_SRS(
   }
 }
 
+// Returns the instruction where the Reg is defined ignoring G_BITCAST.
+static MachineInstr *stripBitCast(Register Reg, MachineRegisterInfo &MRI) {
+  MachineInstr *MI = getDefIgnoringCopies(Reg, MRI);
+  if (MI->getOpcode() == TargetOpcode::G_BITCAST)
+    return getDefIgnoringCopies(MI->getOperand(1).getReg(), MRI);
+  return MI;
+}
+
+// If two 256-bit SRS are input to 512-bit STORE through CONCAT, these are
+// combined into VST.SRS
+bool AIE2InstructionSelector::selectG_AIE_STORE_CONCAT_SRS(
+    MachineInstr &StoreI, MachineRegisterInfo &MRI) {
+
+  Register ConcatResult = (StoreI.uses().begin())->getReg();
+  MachineInstr *ConcatOp = MRI.getUniqueVRegDef(ConcatResult);
+
+  assert(ConcatOp && "Expected SSA.");
+
+  if (getLoadStoreSize(StoreI) != 512 ||
+      ConcatOp->getOpcode() != AIE2::G_INTRINSIC ||
+      cast<GIntrinsic>(*ConcatOp).getIntrinsicID() !=
+          Intrinsic::aie2_concat_I512_I256)
```

Comment:
I'm wondering whether we would want to normalize this to G_CONCAT_VECTORS. Any way, having common pattern recognizer classes for these common glue patterns (e.g. CONCAT(vsz, vsz) -> vszx2)  would really help readability, and be good candidates to have tablegen support for pattern matching.

Reply:
I agree, if we can solve this in PostLegalizer combiner I think the solution can be simpler. Something like:

```
X = CONCAT (SRS(A), SRS(B))
```
To something like:

```
X = SRS2(CONCAT(A, B))
```
And let Selector intact.

Reply:
That indeed seems to be a better way, but I'd rather not split the target-specific handling of intrinsics in one more place, and keep it solely here. I think we all agree the logic for selecting opcodes in this file isn't great, but hopefully we can table-gen-erate most of the logic at some point. Hopefully :)